### ISSUE TITLE
Form requests skip the HTTP middleware stack

### ISSUE THREAD

**Issue by andrewmclagan** [laravel/internals#48]
Im sure there is a good reason, although FormRequests skip the middleware stack.

Currently we have a middleware that transforms any input data coming into our API from snake_case to camelCase.

```
    /**
     * Convert the case of keys on incoming requests.
     *
     * @param  \Dingo\Api\Http\Request $request
     * @param  \Closure  $next
     * @return mixed
     */
    public function handle(Request $request, Closure $next)
    {
        $formatted = CaseTransformer::snakeCaseKeys($request->toArray());

        $request->replace($formatted);

        return $next($request);
    }
```

although when validating Requests with a `FormRequest` those transformations are not present. After investigating the `Illuminate\Foundation\Providers\FoundationServiceProvider` class, its clear that FormRequests are initialised very early on and do not receive any middleware present within the app.

Can someone please explain to me the purpose of form request validators if they are not being transformed through middleware? 

If they are simply to validate raw input then how do we validate input AFTER our HTTP middleware stack? As this is also very important.

thanks!

**Comment by tomschlick** [comment]
I'm not familiar with the internals of the request stack ordering but I can't think of any reason why that might need to come first before middleware. 

Maybe investigate creating a pull request for it if it's not too much work... I'm sure others would find that useful as well to be able to modify the request before it hits the form request/controller.

**Comment by andrewmclagan** [comment]
Oka, although it seems like a futile PR as there is a 99% chance it will get shot down. All I'm after is the reasoning, so we can move forward.

Imagine this scenario:

_You want to add +5 to the year of all age requests (similar to the example in the docs_

What do you do? Simple, you createa middleware:

````
    /**
     * Run the request filter.
     *
     * @param  \Illuminate\Http\Request  $request
     * @param  \Closure  $next
     * @return mixed
     */
    public function handle($request, Closure $next)
    {
        if ($request->input('age')) {
            $request->merge(['age' => $request->input('age') += 5]);
        }

        return $next($request);
    }
````

Although if you want to validate that within a form request, currently you cannot.

**Comment by tomschlick** [comment]
Even if you don't submit the PR, just figuring out how it can be done and posting here might start the discussion about why / how it should or shouldn't be done.

**Comment by andrewmclagan** [comment]
Okay so here goes...

--------
#1
`FormRequest` is initialized with the `Request` object when the `Illuminate\Routing\Events\RouteMatched` event is fired. This event is on line 697 of `Illuminate\Routing\Router`

#2
As you can see no middleware have yet been run, next on line 699:

`$response = $this->runRouteWithinStack($route, $request);` 

the middlewares are run. Thus our `FormRequest` does not receive any middleware.

--------

Basically if we want to achieve this we would need to initialize form requests as a middleware, not seperate from the rest of the HTTP stack.

**Comment by lagbox** [comment]
If the middleware is 'skipped' how do you have access to the session and auth in your FormRequest to do authorization?

**Comment by tomschlick** [comment]
I can't think of any issue with running form requests as a middleware... can you? It might even clean up the logic a bit in that class and allow more flexibility.

**Comment by lagbox** [comment]
Form requests and middleware are completely different things, so I wouldn't know how to answer that.

**Comment by lagbox** [comment]
The FormRequest gets filled from the Request.

I can demonstrate again and again, that what I am doing to the Request in the middleware is showing in my FormRequests.

**Comment by pipegris** [comment]
I have been hitting my head against the wall to know why the hell my FormRequest ignores my middleware request modifications.  I have the same scenario thet @andrewmclagan  shows.

From my controller:
```
public function store(MyCustomRequest $request)
    {
        dd($request->all()); // This shows the raw input
        dd(Request::all()); // This shows the 'processed' input
```
The Facade shows the modifications, while the FormRequest instance doesn'.

Is then the way validating in the controllers or in the middleware the solution? This is dissapointing, since I'm lost here.

**Comment by lagbox** [comment]
Route
```php
Route::resource('test', 'TestController');
```
Controller
```php
...
class TestController extends Controller
{
    public function __construct()
    {
        $this->middleware('testit', ['only' => 'store']);
    }

    public function store(TestRequest $request)
    {
        dd($request->all(), \Request::all());
    }
    ...
}
```
FormRequest
```php
...
class TestRequest extends Request
{
    ...
    public function rules()
    {
        return [];
    }
}
```
Middleware
```php
...
class TestIt
{
    public function handle($request, Closure $next)
    {
        $request->replace(['blah' => 'blah', 'test' => 'test']);
        return $next($request);
    }
}
```

Lets post to that route.
```php
array:2 [▼
  "blah" => "blah"
  "test" => "test"
]

array:2 [▼
  "blah" => "blah"
  "test" => "test"
]
```

Seems like this works exactly as it should.
Lets double check by dumping from inside the FormRequest itself:

```php
class TestRequest extends Request
{
    public function rules()
    {
        dd($this->all());
    }
}
```
Lets post again.
```php
array:2 [▼
  "blah" => "blah"
  "test" => "test"
]
```
Seems to still be working as expected.

Lets triple check by dumping the inputs before replacing them and then from the controller again. (dumping the inputs before replacing and dd'ing from the controller)

```php
class TestIt
{
    public function handle($request, Closure $next)
    {
        dump($request->all());
        $request->replace(['blah' => 'blah', 'test' => 'test']);
        return $next($request);
    }
}
```
Lets post again.
```php
array:4 [▼
  "_token" => "EH89hrlJa2r5kGX7G5c0dYWEqCBFQzQ7LLMoWbVV"
  "name" => "sdfasdf"
  "additional" => "asdfdsf"
  "content" => "asdfsdf"
]

array:2 [▼
  "blah" => "blah"
  "test" => "test"
]

array:2 [▼
  "blah" => "blah"
  "test" => "test"
]
```
Again everything seems to be working as expected.

**Comment by pipegris** [comment]
I made a test with Postman and with PHPStorm's REST test tool. In both cases IT WORKED. However when I try to test my front app (AngularJS), and check the requests made, there's two requests, the preflight OPTIONS, and the subsequent POST (front and back are in different domains). THAT is probably the big difference in our tests.
The clients used to POST to the endpoint are in the same domain, so there's only one request (The POST one). I have this package https://github.com/barryvdh/laravel-cors and by looking a little bit of the code, it handles the OPTION request, and returns the allowed-* headers, the OPTIONS request never get to my middleware/Request/Controller stack. When I try to dump anything from the request I can confirm the OPTIONS request is handled before everything, and all I see is the normal intended request (POST).

I also damp the class of the request ( `Log::info(get_class($this))` in the TestRequest), and if is a single request it shows in every point the same type of instance, but when I make a CORS enabled request, the instance received in the controller (the value of $this in the TestRequest) is the original unmodified request.

I could narrow this by examining the headers sent and received in both tests (with a test client, and with my app) and the problem is when Content-Type: application/json header is on.

**Comment by CJLees01** [comment]
I'm seeing the same result as @andrewmclagan on Laravel 5.3.21 -- it seems that request manipulations via middleware aren't reflected into Form Request classes. 

Here's a route with some custom middleware to run input through a HTML purifier (wraps strings in <p> tags and converts javascript boolean strings into numbers):

```
Route::group(['middleware' => ['purify']], function () {
        Route::post('comment', ['uses' => 'CommentController@store']);
});
```

The middleware is set up in `app/Http/Kernel.php` to point to a `RequestPurifier` class, which in turn manipulates the request (via a job not shown here):

```
class RequestPurifier
{
    use DispatchesJobs;
    public function handle($request, Closure $next)
    {
        $cleaned = $this->dispatch(new PurifyInputJob( $request->all() ));
        $request->replace($cleaned);
        return $next($request);
    }
}
```

I have an `AddCommentRequest` form validator set up to validate the (hopefully purified) input, but if I refer to it in my controller and log the output, I get a non-purified version:

```
public function store(\App\Requests\AddCommentRequest $request)
{
     Log::info($request->all());
}

local.INFO: array (
  'comment' => 'this is my test comment',
  'is_private' => true
) 
```

But keeping everything identical and just changing my controller store function to use the standard `\Illuminate\Http\Request` instead shows the correctly-purified result:

```
public function store(\Illuminate\Http\Request $request, Spec $spec)
{
    Log::info($request->all());
}

local.INFO: array (
  'comment' => '<p>this is my test comment</p>',
  'is_private' => 1
)  

```

It seems that I'm either missing something here, or middleware manipulations aren't reflected into Form Requests.

**Comment by sisve** [comment]
Relevant link. https://github.com/laravel/framework/issues/10791

**Comment by fumpierrez** [comment]
Having this issue on 5.4, do newer versions fix it?
